Common/Util/RateLimit/LeakyBucket.cs

```diff
@@ -0,0 +1,168 @@
+            )
+        {
+        }
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="LeakyBucket"/> class
+        /// </summary>
+        /// <param name="capacity">The maximum number of tokens this bucket can hold</param>
+        /// <param name="sleep">Defines the <see cref="ISleepStrategy"/> used when <see cref="Consume"/> is invoked
+        /// but the bucket does not have enough tokens yet</param>
+        /// <param name="refill">Defines the <see cref="IRefillStrategy"/> that computes how many tokens to add
+        /// back to the bucket each time consumption is attempted</param>
+        /// <param name="timeProvider">Defines the <see cref="ITimeProvider"/> used to enforce timeouts when
+        /// invoking <see cref="Consume"/></param>
+        public LeakyBucket(long capacity, ISleepStrategy sleep, IRefillStrategy refill, ITimeProvider timeProvider = null)
+        {
+            _sleep = sleep;
+            _refill = refill;
+            _capacity = capacity;
+            _available = capacity;
+            _timeProvider = timeProvider ?? RealTimeProvider.Instance;
+        }
+
+        /// <summary>
+        /// Blocks until the specified number of tokens are available for consumption
+        /// and then consumes that number of tokens.
+        /// </summary>
+        /// <param name="tokens">The number of tokens to consume</param>
+        /// <param name="timeout">The maximum amount of time, in milliseconds, to block. An exception is
+        /// throw in the event it takes longer than the stated timeout to consume the requested number
+        /// of tokens</param>
+        public void Consume(long tokens, long timeout = Timeout.Infinite)
+        {
+            if (timeout < Timeout.Infinite)
+            {
+                throw new ArgumentOutOfRangeException(nameof(timeout),
+                    "Invalid timeout. Use -1 for no timeout, 0 for immediate timeout and a positive number " +
+                    "of milliseconds to indicate a timeout. All other values are out of range."
+                );
+            }
+
+            var startTime = _timeProvider.GetUtcNow();
+
+            while (true)
+            {
+                if (TryConsume(tokens))
+                {
+                    break;
+                }
+
+                if (timeout != Timeout.Infinite)
+                {
+                    // determine if the requested timeout has elapsed
+                    var currentTime = _timeProvider.GetUtcNow();
+                    var elapsedMilliseconds = (currentTime - startTime).TotalMilliseconds;
+                    if (elapsedMilliseconds > timeout)
+                    {
+                        throw new TimeoutException("The operation timed out while waiting for the rate limit to be lifted.");
+                    }
+                }
+
+                _sleep.Sleep();
```

Comment:
Not sure whats the use case for this `Consume` implementation, it will block the calling thread until the time tokens are available

Reply:
Leaky bucket is a rate limiting algorithm. So this `Consume` method is _arguably_ the reason the entire class exists.